File: java/CopyListWithRandomPointer.java
import java.util.Map;
import java.util.HashMap;

public class CopyListWithRandomPointer {
    public RandomListNode copyRandomList(RandomListNode head) {
        RandomListNode token = new RandomListNode(0);
        token.next = head;
        head = token;
        RandomListNode copyHead = new RandomListNode(0);
        RandomListNode copyTail = copyHead;
        Map<RandomListNode, RandomListNode> map 
            = new HashMap<RandomListNode, RandomListNode>();
        map.put(null, null);

        for (RandomListNode ptr = head; ptr != null; ptr = ptr.next) {
            if (!map.containsKey(ptr.next)) {
                map.put(ptr.next, new RandomListNode(ptr.next.label));
            }
            if (!map.containsKey(ptr.random) && ptr.random != null) {
                map.put(ptr.random, new RandomListNode(ptr.random.label));
            }
            copyTail.next = map.get(ptr.next);
            copyTail.random = map.get(ptr.random);
            copyTail = copyTail.next;
        }
        return copyHead.next;
    }
}

File: python/copy_list_with_random_pointer.py
#!/usr/bin/env python3

from random_list_node import RandomListNode

class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
        ptr = head
        while ptr != None:
            post = ptr.next
            ptr.next = RandomListNode(ptr.label)
            ptr.next.next = post
            ptr = post
        ptr = head
        while ptr != None:
            if ptr.random != None:
                ptr.next.random = ptr.random.next
            else:
                ptr.next.random = None
            ptr = ptr.next.next
        copy_head = RandomListNode(0)
        copy_tail = copy_head
        ptr = head
        while ptr != None:
            copy_tail.next = ptr.next
            copy_tail = ptr.next
            ptr.next = ptr.next.next
            ptr = ptr.next;
        return copy_head.next

def main():
    solver = Solution()

if __name__ == '__main__':
    main()
